artifacts/artifacts.go

```diff
@@ -16,6 +20,114 @@ import (
 	"github.com/flanksource/incident-commander/api"
 )
 
+type ArtifactContent struct {
+	ActionID string `json:"action_id"`
+	Content  []byte `json:"content"`
+}
+
+// TODO: Memoize this
+func GetArtifactFS(ctx context.Context) (fs.FilesystemRW, *models.Connection, error) {
+	if api.DefaultArtifactConnection == "" {
+		return nil, nil, dutyAPI.Errorf(dutyAPI.ENOTIMPLEMENTED, "no artifact connection configured")
+	}
+
+	conn, err := pkgConnection.Get(ctx, api.DefaultArtifactConnection)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get connection: %w", err)
+	} else if conn == nil {
+		return nil, nil, dutyAPI.Errorf(dutyAPI.ENOTIMPLEMENTED, "the artifacts connection was not found")
+	}
+
+	fs, err := artifacts.GetFSForConnection(ctx, *conn)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get connection: %w", err)
+	}
+
+	return fs, conn, nil
+}
+
+func SaveArtifacts(ctx context.Context, playbookActionID uuid.UUID, generatedArtifacts []artifacts.Artifact) error {
+	if len(generatedArtifacts) == 0 {
+		return nil
+	}
+
+	if api.DefaultArtifactConnection == "" {
+		logger.Warnf("no artifact connection configured")
+		return nil
+	}
+
+	fs, connection, err := GetArtifactFS(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get artifact fs: %w", err)
+	}
+	defer fs.Close()
+
+	for _, a := range generatedArtifacts {
+		a.Path = filepath.Join("playbooks", playbookActionID.String(), a.Path)
+		artifact := models.Artifact{
+			PlaybookRunActionID: utils.Ptr(playbookActionID),
+			ConnectionID:        connection.ID,
+		}
+
+		if err := artifacts.SaveArtifact(ctx, fs, &artifact, a); err != nil {
+			return fmt.Errorf("error saving artifact to db: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func FetchArtifacts(ctx context.Context, maxSize int64, actionIDs ...string) ([]ArtifactContent, error) {
+	if len(actionIDs) == 0 {
+		return nil, nil
+	}
+
+	if api.DefaultArtifactConnection == "" {
+		logger.Warnf("no artifact connection configured")
+		return nil, nil
+	}
+
+	if maxSize == 0 {
+		maxSize = int64(ctx.Properties().Int("max_artifact_size", -1))
+	}
+
+	var actionArtifacts []models.Artifact
+	if err := ctx.DB().Where("playbook_run_action_id = ?", actionIDs).Find(&actionArtifacts).Error; err != nil {
+		return nil, fmt.Errorf("failed to get playbook run action artifacts: %w", err)
+	}
+
+	fs, _, err := GetArtifactFS(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get artifact fs: %w", err)
+	}
+	defer fs.Close()
+
+	var contents []ArtifactContent
+	for _, a := range actionArtifacts {
+		reader, err := fs.Read(ctx, a.Path)
+		if err != nil {
+			return nil, fmt.Errorf("error loading artifact: %w", err)
+		}
+		defer reader.Close()
+
+		if maxSize > 0 {
+			reader = io.NopCloser(io.LimitReader(reader, maxSize))
```

Comment:
I think this should be a tail and not a head (and rename function to TailArtificats)

Reply:
Good point. 

But since artifactFS doesn't natively support this, does it make sense to tail since we'd need to read the entire body and then truncate?

The `LimitReader` here can act as a safeguard that ideally we will never need. The caller can then tail or read all.